File: BOJ/29_유니온파인드/4195_친구네트워크.cpp
#include <iostream>
#include <string>
#include <map>

#define MAXF 200000

using namespace std;

int T,F;
char tmp[21];
string x,y;
pair<int, int> par[MAXF+1];


int find(int x) {
    if(par[x].first == x) return par[x].first;
    else return par[x].first = find(par[x].first);
}

void merge(int x, int y) {
    int p_x = find(x);
    int p_y = find(y);
    if(p_x == p_y) return ;

    par[p_x].first = p_y;
    par[p_y].second += par[p_x].second;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);cout.tie(NULL);
    cin >> T;
    while(T--) {
        cin >> F;
        int cnt = 1;
        map<string,int> names;
        while(F--) {
            cin >> x >> y;
            if(names.find(x) == names.end()) {
                names[x] = cnt++;
                par[names[x]] = make_pair(names[x],1);
            }
            if(names.find(y) == names.end()) { 
                names[y] = cnt++;
                par[names[y]] = make_pair(names[y],1);
            }
            merge(names[x],names[y]);

            cout << par[find(names[y])].second << "\n";
        }
    }


}

File: BOJ/29_유니온파인드/4195_친구네트워크.py
import sys

read = sys.stdin.readline

def find(x) :
    if par[x][0] == x : return par[x][0]
    else :
        par[x][0] = find(par[x][0])
        return par[x][0]

def merge(x,y) :
    p_x = find(x)
    p_y = find(y)

    if p_x == p_y : return

    par[p_x][0] = p_y
    par[p_y][1] += par[p_x][1]

T = int(read())
for _ in range(T) :
    F = int(read())

    par = {}
    cnt = 1
    for _ in range(F) :
        x,y = read().split()
        
        if x not in par : 
            par[x] = [x,1]
        if y not in par :
            par[y] = [y,1]

        merge(x,y)
        
        p_y = find(y)
        print(par[p_y][1])

#def 시간초과난 코드 :
    # T = int(read())
    # for _ in range(T) :
    #     F = int(read())

    #     par = {}
    #     cnt = 1
    #     for _ in range(F) :
    #         x,y = read().split()
            
    #         if x not in par : 
    #             par[x] = x
    #         if y not in par :
    #             par[y] = y

    #         merge(x,y)
    #         net = []
    #         for idx,val in par.items() :
    #             if idx != x and find(x) == find(idx) and idx not in net :
    #                 net.append(idx)
    #             if idx != y and find(y) == find(idx) and idx not in net :
    #                 net.append(idx)
            
    #         print(len(net))